tests/test_2306_cppyy_git.py

```diff
@@ -0,0 +1,202 @@
+# __castcpp__
+# PyObject* __castcpp__(PyObject*)
+
+
+@pytest.mark.skip("invalid range expression of type 'awkward::ArrayView'")
+def test_array_base_git():
+    array = ak.Array(
+        [
+            [{"x": 1, "y": [1.1]}, {"x": 2, "y": [2.2, 0.2]}],
+            [],
+            [{"x": 3, "y": [3.0, 0.3, 3.3]}],
+        ]
+    )
+
+    generator = ak._connect.cling.togenerator(array.layout.form, flatlist_as_rvec=False)
+    lookup = ak._lookup.Lookup(array.layout)  # noqa: F841
+
+    # subsequent calls to the generator do not change the generated type
+    print("test line 114:", generator.dataset())
+    print("test line 115:", generator.dataset())
+
+    # Note, if 'awkward_array' passed in by-value,
+    # it will be slides to the base class
+    # (if T is indeed a base class).
+    # Pass it by pointer or reference if
+    # the derived class has extra state.
+    source_code_cpp = """
+    template<typename T>
+    double go_fast_cpp_1(T awkward_array, ssize_t length) {
+        double out = 0.0;
+
+        // FIXME: ArrayView base class does not provide begin() and end()
+        // nor it provides a subscript operator.
+        //
+        // for (auto list : awkward_array) {
+        //   for (auto record : list) {
+        //
+        // for (int64_t i = 0; i < length; i++) {
+        //   for (auto record : awkward_array[i]) {
+        //
+
+            for (auto list : awkward_array) {
+              for (auto record : list) {
+                for (auto item : record.y()) {
+                    out += item;
+                }
+            }
+        }
+
+        return out;
+    }
+    """
+
+    cppyy.set_debug()
+    cppyy.cppdef(source_code_cpp)
+    out = cppyy.gbl.go_fast_cpp_1["awkward::ArrayView"](array, len(array))
+    print("test line 152:", out)
+
+
+# On the Python side:
+#     * make a new buffer (full of pointers)
+#
+# On the C++ side:
+#     * new type named ArrayView_aslkdiuhflskeudfhn(buffer, size)
+
+
+def test_array_derived_git():
+    array = ak.Array(
+        [
+            [{"x": 1, "y": [1.1]}, {"x": 2, "y": [2.2, 0.2]}],
+            [],
+            [{"x": 3, "y": [3.0, 0.3, 3.3]}],
+        ]
+    )
+
+    # Note, if 'awkward_array' passed in by-value,
+    # it will be slides to the base class
+    # (if T is indeed a base class).
+    # Pass it by pointer or reference if
+    # the derived class has extra state.
+    source_code_cpp = """
+    template<typename T>
+    double go_fast_cpp_2(T awkward_array) {
+        double out = 0.0;
+
+        for (auto list : awkward_array) {
+            for (auto record : list) {
+                for (auto item : record.y()) {
+                    out += item;
+                }
+            }
+        }
+
+        return out;
+    }
+    """
+
+    cppyy.set_debug()
+    cppyy.cppdef(source_code_cpp)
+
+    # FIXME: a tuple of parameters: move it to __cpptype__??
+    def convert(array):
+        return (0, len(array), 0, array._lookup.arrayptrs, 0)
```

Comment:
These are the parameters needed to generate this array dataset. It corresponds to a `generator.dataset()` output: `awkward::ListArray_x9C5rj3ig8(0, length, 0, reinterpret_cast<ssize_t*>(ptrs), 0)`

Reply:
the it's simply:
```python
out = cppyy.gbl.go_fast_cpp_2[f"awkward::{array.cpptype}"](convert(array))
```